### PR TITLE
Implement standardized "use flags" for Nixpkgs

### PR DESCRIPTION
**Pull request by matthewbauer**
This is something I worked up. I'm interested whether people think this is a good way to go. It basically uses the "overlay" system to implement standardized config options in Nixpkgs. To use them in your packages, you must have an argument with the same name as one of the standardized options. For example,

```
{ stdenv, pkgconfig, fetchurl
, withPulseAudio, libpulseaudio
, withAlsa, alsaLib
}:
```

withPulseAudio and withAlsa will automatically be set to the appropriate value for your platform. You can override it by setting the config "withPulseAudio" or "withAlsa" to true.

```
nix-build -A foo --arg config '{ withPulseAudio = false; }'
```

This is basically what was discussed in #12877. The new overlay system makes this extremely straightforward. For this to work well, people need to be aware of it and use it where appropriate. I want to generate some documentation to go with this.

Are there other options that should be included?

### CODE REVIEW

**Comment by xeji** [comment]
Nice! I like the idea of some standardized build options. Some initial thoughts:

- It would have to be well documented.
- I'm not sure of is how to get consensus on the global default values. They are important because they can influence what is built on Hydra and available in the binary cache.
- Whats the interaction with default values in a package expresson? For instance, what happens if the global option `withGTK=false` but the `qemu` expression has `withGTK ? true` ? IMO defaults in the package should have priority over global options but I'm not sure your implementation does that.

**Comment by matthewbauer** [comment]
> Whats the interaction with default values in a package expresson? For instance, what happens if the global option withGTK=false but the qemu expression has withGTK ? true ? IMO defaults in the package should have priority over global option defaults but I'm not sure your implementation does that.

So the global default will always be used. In fact it probably makes sense to not even provide a package default for any of the listed configss. Right now very few packages are using a "with*" so hopefully it will not lead to anything unexpected.

File: pkgs/top-level/options.nix [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    default = true;
+    type = types.bool;
+    description = ''
+
+      Whether to compile packages with support for the Qt graphical toolkit
+
+    '';
+  };
+
+  withGTK = mkOption {
```
[the remark this answers is not in the document]
Comment (matthewbauer) [reply pkgs/top-level/options.nix]: We can do that although I'm not quite sure if that's better. GTK is usually capitilized.

**Comment by xeji** [comment]
Ok, so we might have things like `withGTK` to use the global option and another name `gtkSupport ? true` if we think the package should user a different default. That will work (but may create a little confusion).

File: pkgs/top-level/options.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (jtojnar) [reply pkgs/top-level/options.nix]: So is ALSA. See, for example, https://stackoverflow.com/questions/15526107/acronyms-in-camelcase for question about acronyms in identifiers. I do not really care which style we choose but the inconsistency is jarring.

[the remark this answers is not in the document]
Comment (Anton-Latukha) [reply pkgs/top-level/options.nix]: Yes. that only first letter capitalizes also bothered me in `camelCase`.
But it is a trade-off for `thisHTTPUIFFMEGNamesBeingReadable`.

https://github.com/NixOS/nixpkgs/search?utf8=%E2%9C%93&q=withgtk+OR+usegtk+OR+withqt+OR+useqt+OR+withalsa+OR+usealsa&type=

File: pkgs/top-level/options.nix [hunk 2]
```diff
@@ -0,0 +1,128 @@
+{lib, stdenv}:
```
Comment (Ericson2314) [inline pkgs/top-level/options.nix]: Can you use `hostPlatform` instead?

**Comment by Ericson2314** [comment]
@edolstra has talked about at various points trying to avoid non-packages on the top level, all things equal. I'm somewhat sympathetic---and while various functions are sort of ad-hoc exception, use flags would be a whole class of non-packages on the top level. Maybe we could put these all in `options.*`, and to a trick [like we do with `xorg` in `splice.nix`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/splice.nix#L39) so `callPackage` sees them?

**Comment by matthewbauer** [comment]
> @edolstra has talked about at various points trying to avoid non-packages on the top level, all things equal. I'm somewhat sympathetic---and while various functions are sort of ad-hoc exception, use flags would be a whole class of non-packages on the top level. Maybe we could put these all in options.*, and to a trick like we do with xorg in splice.nix so callPackage sees them?

Yeah that would probably work! We could also have each package use the "config" attribute directly. The main reason I like the original though is it's pretty easy for maintainers to grok how to handle it (in fact many are already using them).

File: pkgs/top-level/options.nix (hunk 2 above)
Comment (matthewbauer) [reply pkgs/top-level/options.nix]: Yeah definitely! targetPlatform might actually make more sense though. In fact, one of the big reasons for this is getting rid of a lot of the ```stdenv.isDarwin``` and ```stdenv.isLinux``` that is spread accross Nixpkgs. Quite a few should be easily replaced with ```withApple```, ```withSystemd``` and ```withAlsa```.

**Comment by Ericson2314** [comment]
Yeah I don't really care. The names make pretty clear they aren't packages.

**Comment by maurer** [comment]
Caveat: I'm not some kind of nix decision maker, just leaving my 2 cents.

Technically / in terms of the code, this seems fine, but I think there are some policy/development strategy questions we need to answer before we mainline a change like this:

* Currently, most packages have one configuration. There are exceptions, but usually if a package builds and runs on hydra, we know it's put together correctly. How are we going to approach testing changes if there's an explosion of configuration options?
Having a robust testing infrastructure in Hydra is one of the reasons we are able to get by with so few developers and still remain reasonably up to date.
I am concerned that adding a wide variety of untested configurations would end up generating a large number of (valid) bug reports that need to be triaged and fixed.
Approaches I can see are:
  * Only test the "defaults", e.g. with the `options' record empty.
  * Test every global set of options which is "compatible" (I think this is computationally implausible)
  * Have some set of endorsed `options' records which are tested
* If we permit different options for a package than the global setting, is it broken/incorrect behavior for a package to fail in that case? For example, if we had a `withKrb' flag for Kerberos support, and I have it on globally, but off for a package via an override, it could receive a header from the dependency which referenced a Kerberos header. Depending on the build system, this could do any of:
  * Error out, if it's not using `pkg-config' or similar to inherit search paths
  * End up linking Kerberos in anyways, but not have support on the interface layer
  * End up with full Kerberos support
Are any of these allowed? All of them? How do we deal with this?
* Which packages will be present in the NixOS cache? The Gentoo USEflags approach largely makes sense because everyone is going to be building locally on a heterogeneous background anyways, so configuration explosion was always a problem. In our case, most users use the cache for most packages, manually building only the more esoteric packages or ones they need bleeding edge versions of.
This is not necessarily a problem that needs a complete solution, but again, we'd need some kind of policy on which versions of packages would be cached.

Overall, I like the idea, and I miss `USE' from Gentoo. However, I would hope that in getting the flags ourselves, we don't inherit Gentoo's testing difficulties.

**Comment by matthewbauer** [comment]
> Currently, most packages have one configuration. There are exceptions, but usually if a package builds and runs on hydra, we know it's put together correctly. How are we going to approach testing changes if there's an explosion of configuration options?

Most likely with some sort of profiles. This is basically what we do with cross now: there are tons of possible config system options, but we just provide a couple of "examples" in https://github.com/NixOS/nixpkgs/blob/master/lib/systems/examples.nix.

I would like to get something like that and a release-options.nix file that build a few combinations of configuration (along with the defaults of course). Some possible ones:

- headlessLinux - all X11 and stuff disabled
- freeDarwin - all proprietary Apple SDK removed
- waylandLinux - Linux built with Wayland APIs
- staticLinux - everything statically compiled
- staticDarwin - everything statically compiled
- slnosLinux - Linux without Systemd or Pulseaudio

Note that lots of packages are going to be the same between these profiles & not a ton of rebuilding will be needed. I'm hoping with time we can get better at structuring dependencies so there's a lot of shared infrastructure between them (this will improve other things as well). These examples don't all have to be built by Hydra yet.

**Comment by Anton-Latukha** [comment]
My idea is that plethora of configuration options can be integrated only if NixOS+IPFS ever takes into reality.
Hydra needs to build main packages, custom option packages versions need to be build and tested by users, and info shared to testing system through IPFS. Something like that.
So if someone builds package with custom options, it get shared through IPFS, as also it's log to the central CI.

**Comment by matthewbauer** [comment]
>So if someone builds package with custom options, it get shared through IPFS, as also it's log to the central CI.

I don't know how you could ever trust that IPFS has valid binaries. Without the binary how would you know that the hash in IPFS is correct? It's definitely true we need more builders for Nixpkgs for this to work well - but I think we can improve this without ridiculous costs. We may need to constrain it to a subset of packages for some profiles - things like SDL games should be marked as "broken" when you're headless for example.

**Comment by Anton-Latukha** [comment]
Yes, verifying builds is the question I deliberately omitted.

Because I was not talking about package cache, but about build and test results to be shared to CI.

So maintainers can see those logs, and use that information.

**Comment by copumpkin** [comment]
If you recall from my original issue, I was extremely wary of getting too many of these options. I basically proposed very high-level "headless" ones and nothing else, because the chances of arbitrary combinations (with exponential possibilities) working is pretty low IMO, and will cause massive rebuilds where people will only realize they don't work after waiting 5 hours.

Anyway, I like that you're doing this, but I'd be very careful and considered about which flags we want to put in there. I could imagine one or two maybe being supported nicely, but too many more and folks will just lose faith in the mechanism after it breaks too often.

**Comment by matthewbauer** [comment]
I guess what I mostly dislike with “headless” is that it’s pretty vague. There’s not really a headless configure flag - usually you just need to disable all of the GUI backends. My goal with these options is that they will make sense for packages to use without having to make a judgement call on what it means to be headless. I think a headless “profile” makes sense though.

**Comment by oxij** [comment]
Also forgot to mention. I'd love `headlessLinux` to be `nographicsLinux // nosoundLinux` or similar. I find those things to be useful by themselves too, e.g. a X11-less singe board computer (RPi, BBB) that has ALSA enabled and is connected to a USB sound card and a speaker set to make a standalone music player/announcement system/personal assistant (when connected to a mic).

**Comment by matthewbauer** [comment]
Closing for now. I think custom overlays are the way to go for this.

**Pull request closed without merging**